### PR TITLE
optimize(Path2D): replace reassign-concat with for-push

### PR DESCRIPTION
**Pull request by milahu**

### CODE REVIEW

**Comment by jtenner** [comment]
This might be a valid performance gain, with the trade-off that the software becomes less simple. Also, why did code coverage decrease?

**Comment by jtenner** [comment]
@milahu I think this might be another micro-optimization, however, it's probably fine.

https://coveralls.io/builds/35861728/source?filename=src%2Fclasses%2FPath2D.js#L43

Please address this code coverage issue if you can. We need to add a way to validate that push gets called.

**Comment by milahu** [comment]
sorry .. note to self: run tests offline

File: src/classes/Path2D.js [hunk 1]
```diff
@@ -39,6 +39,7 @@ export default class Path2D {
       throw new TypeError(
         "Failed to execute 'addPath' on 'Path2D': parameter 1 is not of type 'Path2D'."
       );
-    this._path = this._path.concat(path._path);
+    for (let i = 0; i < path._path.length; i++)
+      this._path.push(path._path[i]);
```
Comment (hustcc) [inline src/classes/Path2D.js:43]: It doesn't seem to make any difference?

Comment (milahu) [reply src/classes/Path2D.js:43]: its faster : )

Comment (hustcc) [reply src/classes/Path2D.js:43]: I think we do not consider the perf of language style, compile will do it.^_^

Comment (milahu) [reply src/classes/Path2D.js:43]: you mean babel should optimize this? bad news:

```js
// jest-canvas-mock/lib/classes/Path2D.js
    this._path = this._path.concat(path._path);
```

its a micro optimization, but then why not

Comment (hustcc) [reply src/classes/Path2D.js:43]: `concat` is more simple for code.

Comment (milahu) [reply src/classes/Path2D.js:43]:
```js
Array.prototype._pushArray = function _pushArray(a) {
  for (let i = 0; i < a.length; i++)
    this.push(a[i]);
  //return this;
};

this._path._pushArray(path._path);
```

`this._path.push(...path._path);` would be limited by call stack size

Comment (hustcc) [reply src/classes/Path2D.js:43]: I still think this is what the compiler should do, and the upper code should not deal with it.

Comment (hustcc) [reply src/classes/Path2D.js:43]: image

many `concat` usage.

Comment (milahu) [reply src/classes/Path2D.js:43]: > I still think this is what the compiler should do

compiler? the babel compiler or the JS runtime?
some solutions are generally slower (concat + reassign, functional loops, immutable data) 
and are hard to optimize in the runtime

the pragmatic reason for "no" is: its just a micro optimization, with almost zero gain

Comment (hustcc) [reply src/classes/Path2D.js:43]: Sorry, I mean v8, not babel.

Comment (hustcc) [reply src/classes/Path2D.js:43]: image

**Pull request merged**